src/hotspot/share/code/codeCache.cpp

```diff
@@ -604,6 +604,7 @@ void CodeCache::free(CodeBlob* cb) {
     heap->set_adapter_count(heap->adapter_count() - 1);
   }
 
+  cb->~CodeBlob();
   // Get heap for given CodeBlob and deallocate
   get_code_heap(cb)->deallocate(cb);
```

Comment:
I am not sure it look correct - `cb` is used after you call destructor.
I think it works because destructor only has assert but I don't think it is right thing to do here.

Reply:
`get_code_heap` and `deallocate` only use the address as a chunk of memory, but not as a `CodeBlob`.
Maybe the parameter to `get_code_heap` should be made `void*` too to make it clear that what's passed here is an opaque address?
`deallocate` unwraps the given pointer to a `HeapBlock` immediately, not treating it as a `CodeBlob` any more.
So I think it is the appropriate location to call the destructor.

These two lines resemble a `delete` call: first call the destructor, then pass the pointer to `malloc`.

Reply:
>  Maybe the parameter to get_code_heap should be made void* too to make it clear that what's passed here is an opaque address?

I prototyped this in the latest commit.